### PR TITLE
Feature/issue-3107 add beta negative binomial lpmf

### PR DESCRIPTION
**Pull request by lingium**
## Summary

With this PR the lpmf of beta negative binomial distribution are added.
See issue https://github.com/stan-dev/math/issues/3107

Expressions involved:

PMF

$$f(n|r, \alpha ,\beta)={\frac {\Gamma (r+n)}{n!\Gamma (r)}}{\frac {\Gamma (\alpha +r)\Gamma (\beta +n)\Gamma (\alpha +\beta )}{\Gamma (\alpha +r+\beta +n)\Gamma (\alpha )\Gamma (\beta )}}$$

Partial derivatives w.r.t the three parameters

$$
\begin{aligned}
\frac{\partial \log f}{\partial r} &= \psi(n+r) - \psi(n+r+\alpha+\beta) - \psi(r) + \psi(r+\alpha) \\
\frac{\partial \log f}{\partial \alpha} &= \psi(\alpha+\beta) - \psi(n+r+\alpha+\beta) - \psi(\alpha) + \psi(r+\alpha) \\
\frac{\partial \log f}{\partial \beta} &= \psi(\alpha+\beta) - \psi(n+r+\alpha+\beta) + \psi(n+\beta) - \psi(\beta)
\end{aligned}
$$

## Tests
Test is written follow the [guide](https://mc-stan.org/math/md_doxygen_2contributor__help__pages_2adding__new__distributions.html). 

## Side Effects
No.

## Release notes
`beta_neg_binomial_lpmf` is available if merged. Allows modeling heavy-tail count data.

## Checklist

- [x] Copyright holder: Zhi Ling

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Coding-Style-and-Idioms) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

File: stan/math/prim/prob/beta_neg_binomial_lpmf.hpp
Comment (SteveBronder) [inline stan/math/prim/prob/beta_neg_binomial_lpmf.hpp:208]: Can you move all of the for loops you have separated in the above into this for loop? This would be nice because it would cut down on a bit of the code and calculations you have to do. For instance `digamma_n_r` and `digamma_r` are only used when `T_r` is an autodiff type. So you can make this like 

```c++
    if (!is_constant_all<T_r>::value) {
      partials<0>(ops_partials)[i] += digamma(n_vec[i] + r_vec.val(i)) - digamma_n_r_alpha_beta[i]
                                      - (digamma(r_vec.val(i)) - digamma_r_alpha[i]);
    }
```

And you only need to calculate `digamma_r_alpha` if `T_r` or `T_alpha` is not a constant

**Review by SteveBronder — CHANGES_REQUESTED** [review]

**Comment by WardBrian** [comment]
@lingium not a review itself, but wanted to chime in to say super nice work so far! It's no small feat to have a PR pass CI on the first try; I hardly do myself even now.

Nice meeting you at StanCon, and glad you're contributing to Stan!

**Comment by lingium** [comment]
> @lingium not a review itself, but wanted to chime in to say super nice work so far! It's no small feat to have a PR pass CI on the first try; I hardly do myself even now.
> 
> Nice meeting you at StanCon, and glad you're contributing to Stan!

Hey @WardBrian, thanks for your support! It was a pleasure meeting you at StanCon as well. I'm glad to be contributing to Stan :)

**Comment by lingium** [comment]
Seems like I forgot to pull pre-commit hook's commit to local

**Comment by lingium** [comment]
May have to require your approval to one of the workflows to continue

File: stan/math/prim/prob/beta_neg_binomial_lpmf.hpp
Comment (SteveBronder) [reply stan/math/prim/prob/beta_neg_binomial_lpmf.hpp]: Changes look great! Going to rubber duck this on Monday then should be good!

**Review by SteveBronder — APPROVED** [review]
Looks good to me! Will github let you press the merge button? If not I can, though we use to be able to let contributors click the button themselves which imo feels nice

**Comment by lingium** [comment]
Thanks for the review @SteveBronder! From my side it says merging is blocked, so please go ahead from your end

**Comment by SteveBronder** [comment]
@lingium check your email. You should have received an invite to the stan org and with that I believe you will be able to click the merge button :)

**Comment by lingium** [comment]
Hi @SteveBronder, got the invite. I was tagged as a collaborator now but don't seem to have merge pr permission
image

**Comment by WardBrian** [comment]
@lingium maybe now?

**Comment by lingium** [comment]
Yes, it's working now!

**Comment by WardBrian** [comment]
@lingium are you also interested in adding this to the compiler yourself, or would you prefer another developer do that part? It's easy enough for me to do if you want to provide a list of signatures you'd expect support for

**Comment by lingium** [comment]
@WardBrian could you help me to add this time? Cuz it's super easy for you. Here's list of signatures
```
real beta_neg_binomial_lpmf(int n, real r, real alpha, real beta);
real beta_neg_binomial_lpmf(int n, vector r, real alpha, real beta);
real beta_neg_binomial_lpmf(int n, real r, vector alpha, real beta);
real beta_neg_binomial_lpmf(int n, real r, real alpha, vector beta);
real beta_neg_binomial_lpmf(int n, vector r, vector alpha, real beta);
real beta_neg_binomial_lpmf(int n, vector r, real alpha, vector beta);
real beta_neg_binomial_lpmf(int n, real r, vector alpha, vector beta);
real beta_neg_binomial_lpmf(int n, vector r, vector alpha, vector beta);
real beta_neg_binomial_lpmf(array[] int n, real r, real alpha, real beta);
real beta_neg_binomial_lpmf(array[] int n, vector r, real alpha, real beta);
real beta_neg_binomial_lpmf(array[] int n, real r, vector alpha, real beta);
real beta_neg_binomial_lpmf(array[] int n, real r, real alpha, vector beta);
real beta_neg_binomial_lpmf(array[] int n, vector r, vector alpha, real beta);
real beta_neg_binomial_lpmf(array[] int n, vector r, real alpha, vector beta);
real beta_neg_binomial_lpmf(array[] int n, real r, vector alpha, vector beta);
real beta_neg_binomial_lpmf(array[] int n, vector r, vector alpha, vector beta);
```
Another thing is, we would prefer small PRs, right? If so for the cdf, lcdf, lccdf, and rng, I will submit in different PRs.

**Comment by WardBrian** [comment]
I’ll take a look at that today!

And yes, it is easiest if they are each in their own PR. If I remember correctly from Stan Con, isn’t there also an alternative parameterization you are interested in?

**Comment by lingium** [comment]
Thanks Brian! 

Yes, there is an alternative parameterization. It becomes `neg_binomial_2` when one of the parameters is equal to 0. Except for that, I'm planning Waring and Yule distribution, both are special cases of beta negative binomial.

**Pull request merged**
